File: python/14_longest_common_prefix.py
def longest_common_prefix(strs:[str]):
    res = ""
    for i in range(len(strs[0])):
        for s in strs:
            if i == len(s) or s[i] != strs[0][i]:
                return res
        res += strs[0][i]
    return res

def longest_common_prefix_1(strs:[str]):
    res = ""
    for chars in zip(*strs):
        if all(c == chars[0] for c in chars):
            res += chars[0]
        else:
            break
    return res


def longest_common_prefix_2(strs:[str]):
    if not strs:
        return ""
    prefix = strs[0]
    for s in strs[1:]:
        while not s.startswith(prefix):
            prefix = prefix[:-1]
    return prefix

File: typescript/14-Longest-Common-Prefix.ts
// First solution
function longestCommonPrefix(strs: string[]): string {
  const prefixes: string[] = [];
  for (let i = 0; i < strs[0].length; i++) {
    prefixes.push(strs[0].slice(0, i + 1))
  }

  for (let i = prefixes.length - 1; i >= 0; i--) {
    if (strs.every(str => str.startsWith(prefixes[i]))) {
      return prefixes[i]
    }
  }

  return '';
};

// Second solution
function longestCommonPrefix2(strs: string[]): string {
  strs.sort();
  const firstItem = strs[0];
  const lastItem = strs[strs.length - 1];
  let result = '';

  for (let i = 0; i < firstItem.length; i++) {
    if (firstItem[i] !== lastItem[i]) {
      break;
    }

    result += firstItem[i];
  }

  return result;
};
